packages/spatial/src/input/systems/ClientInputSystem.tsx

```diff
@@ -188,87 +205,138 @@ const execute = () => {
-        if (
-          !layerHit ||
-          !layerHit.intersection.object.visible ||
-          (layerHit.intersection.object as Mesh<any, MeshBasicMaterial>).material?.opacity < 0.01
-        )
-          continue
-        intersectionData.push({ entity, distance: layerHit.intersection.distance })
-      }
 
-      const physicsWorld = getState(PhysicsState).physicsWorld
+        const physicsWorld = getState(PhysicsState).physicsWorld
 
-      // 2nd heuristic is physics colliders
-      if (physicsWorld) {
-        const hits = Physics.castRay(physicsWorld, inputRaycast)
-        // console.log("objDir inputraycast direction = " + inputRaycast.direction.toArray())
-        for (const hit of hits) {
-          if (!hit.entity || !hasComponent(hit.entity, InputComponent)) continue
-          intersectionData.push({ entity: hit.entity, distance: hit.distance })
+        // 2nd heuristic is physics colliders
+        if (physicsWorld) {
+          const hits = Physics.castRay(physicsWorld, inputRaycast)
+          for (const hit of hits) {
+            if (!hit.entity || !hasComponent(hit.entity, InputComponent)) continue
+            intersectionData.push({ entity: hit.entity, distance: hit.distance })
+          }
         }
-      }
 
-      // 3rd heuristic is bboxes
-      for (const entity of inputBoundingBoxes()) {
-        const boundingBox = getComponent(entity, BoundingBoxComponent)
-        const hit = inputRay.intersectBox(boundingBox.box, bboxHitTarget)
-        if (hit) {
-          intersectionData.push({ entity, distance: inputRay.origin.distanceTo(bboxHitTarget) })
+        // 3rd heuristic is bboxes
+        for (const entity of inputBoundingBoxes()) {
+          const boundingBox = getComponent(entity, BoundingBoxComponent)
+          const hit = inputRay.intersectBox(boundingBox.box, bboxHitTarget)
+          if (hit) {
+            intersectionData.push({ entity, distance: inputRay.origin.distanceTo(bboxHitTarget) })
+          }
         }
       }
     }
 
     const sortedIntersections = intersectionData.sort((a, b) => a.distance - b.distance)
-
     const sourceState = getMutableComponent(sourceEid, InputSourceComponent)
-    sourceState.intersections.set(sortedIntersections)
 
-    const capturedEntity = getState(InputState).capturingEntity
+    //TODO check all inputSources sorted by distance list of InputComponents from query, probably similar to the spatialInputQuery
+    //Proximity check ONLY if we have no raycast results, as it is always lower priority
+    if (
+      capturedEntity.value === UndefinedEntity &&
+      sortedIntersections.length === 0 &&
+      !hasComponent(sourceEid, InputPointerComponent)
+    ) {
+      let closestEntity = UndefinedEntity
+      let closestDistanceSquared = Infinity
+
+      //use sourceEid if controller (one InputSource per controller), otherwise use avatar rather than InputSource-emulated-pointer
+      const selfAvatarEntity = AvatarComponent.getSelfAvatarEntity()
```

Comment:
again, AvatarComponent must not be referenced in the spatial package

Reply:
removed references to AvatarComponent